hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java

```
@@ -57,43 +59,88 @@ public ContractCallService(final MirrorEvmTxProcessorFacade mirrorEvmTxProcessor
         gasPerSecondMetricMap = Collections.unmodifiableMap(gasPerSecondMetricEnumMap);
     }
 
-    public String processCall(final CallServiceParameters body) {
-        final var txnResult = doProcessCall(body);
+    public String processCall(final CallServiceParameters params) {
+        if (params.isEstimate()) {
+            return estimateGas(params);
+        }
+
+        final var ethCallTxnResult = processEthTxn(params);
+        validateTxnResult(ethCallTxnResult, params.getCallType());
 
-        final var callResult = txnResult.getOutput() != null
-                ? txnResult.getOutput() : Bytes.EMPTY;
+        final var callResult = ethCallTxnResult.getOutput() != null
+                ? ethCallTxnResult.getOutput() : Bytes.EMPTY;
 
         return callResult.toHexString();
     }
 
-    private HederaEvmTransactionProcessingResult doProcessCall(final CallServiceParameters body) {
+    private String estimateGas(final CallServiceParameters params) {
+        final var providedGasLimit = params.getGas();
+        long lo = 0;
+        long hi = providedGasLimit;
+
+        int minDiffBetweenIterations = 1200;
+        long prevGasLimit = providedGasLimit;
+        HederaEvmTransactionProcessingResult txnResult;
+        do {
+            // Take a guess at the gas, and check transaction validity
+            long mid = (hi + lo) / 2;
+            params.setGas(mid);
+            txnResult = processEthTxn(params);
+
+            boolean err = !txnResult.isSuccessful() || txnResult.getGasUsed() < 0;
```

Why would we keep retrying if the transaction is unsuccessful for any reason besides insufficient gas? If the payload had invalid input or if an entity/state is missing corrupt, or any other unrecoverable error we should fail fast and exit the loop. Or do these type of errors bubble up as uncaught exceptions?